presto-testng-services/src/main/java/io/prestosql/testng/services/FlakyTestRetryAnalyzer.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.testng.services;
+
+import io.airlift.log.Logger;
+import org.testng.IRetryAnalyzer;
+import org.testng.ITestNGMethod;
+import org.testng.ITestResult;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.lang.String.format;
+
+public class FlakyTestRetryAnalyzer
+        implements IRetryAnalyzer
+{
+    private static final Logger log = Logger.get(FlakyTestRetryAnalyzer.class);
+
+    public static final int ALLOWED_RETRIES_COUNT = 2;
+
+    @GuardedBy("this")
+    private final Map<String, Long> retryCounter = new HashMap<>();
+
+    @Override
+    public boolean retry(ITestResult result)
+    {
+        if (!isTestRetryable(result)) {
+            return false;
+        }
+        long retryCount;
+        synchronized (this) {
+            String name = getName(result.getMethod());
```

Comment:
Does this work for parameterized methods (dataprovider)?
maybe we could explicitly detect & reject this?